File: C++/5. 1758. Minimum Changes to Make Alternating Binary String.cpp
#include <iostream>
#include <string>

int count_changes(std::string s, char start_char) {
    int changes = 0;
    char curr_char = start_char;
    for (int i = 0; i < s.length(); i++) {
        if (s[i] == curr_char) {
            changes += 1;
        }
        curr_char = '0' + '1' - curr_char;
    }
    return changes;
}

class Solution {
public:
    int minOperations(std::string s) {
        return std::min(count_changes(s, '0'), count_changes(s, '1'));
    }
};

int main() {
    Solution sol;
    std::string s = "0100";
    std::cout << sol.minOperations(s) << std::endl; // Output: 1

    s = "10";
    std::cout << sol.minOperations(s) << std::endl; // Output: 0

    s = "1111";
    std::cout << sol.minOperations(s) << std::endl; // Output: 2

    return 0;
}

File: Python/5. 1758. Minimum Changes To Make Alternating Binary String.py
class Solution:
    def minOperations(self, s: str) -> int:
        # Create a helper function to check for
        # every starting charachter amongst '0' and '1'
        def count_changes(start_char):
            changes = 0
            curr_char = start_char
            for i in range(len(s)):
                if(s[i] == curr_char):
                    changes += 1
                curr_char = '0' if curr_char == '1' else '1'
            return changes
        return min(count_changes('0'), count_changes('1'))
s = "0100"
print(minOperations(s)) # Output: 1

s = "10"
print(minOperations(s)) # Output: 0

s = "1111"
print(minOperations(s)) # Output: 2
